src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2773,24 +2789,51 @@ void ProcessRequiresUnreferencedCode (MethodDefinition method, DependencyKind de
 				return;
 
 			default:
-				// DirectCall, VirtualCall and NewObj are handled by ReflectionMethodBodyScanner
-				// This is necessary since the ReflectionMethodBodyScanner has intrinsic handling for some
-				// of the annotated methods annotated (for example Type.GetType)
-				// and it knows when it's OK and when it needs a warning. In this place we don't know
-				// and would have to warn every time
-
 				// All other cases have the potential of us missing a warning if we don't report it
 				// It is possible that in some cases we may report the same warning twice, but that's better than not reporting it.
 				break;
 			}
 
-			// All override methods should have the same annotations as their base methods (else we will produce warning IL2046.)
-			// When marking override methods with RequiresUnreferencedCode on a type annotated with DynamicallyAccessedMembers,
-			// we should only issue a warning for the base method.
-			if (dependencyKind != DependencyKind.DynamicallyAccessedMember ||
-				!method.IsVirtual ||
-				Annotations.GetBaseMethods (method) == null)
-				CheckAndReportRequiresUnreferencedCode (method);
+			// All override methods should have the same annotations as their base methods
+			// (else we will produce warning IL2046 or IL2092 or some other warning).
+			// When marking override methods via DynamicallyAccessedMembers, we should only issue a warning for the base method.
+			if (dependencyKind == DependencyKind.DynamicallyAccessedMember &&
+				method.IsVirtual &&
+				Annotations.GetBaseMethods (method) != null)
+				return;
+
+			CheckAndReportRequiresUnreferencedCode (method);
+
+			if (_context.Annotations.FlowAnnotations.ShouldWarnWhenAccessedForReflection (method)) {
+				// If the current scope has analysis warnings suppressed, don't generate any
+				if (ShouldSuppressAnalysisWarningsForRequiresUnreferencedCode ())
+					return;
+
+				// ReflectionMethodBodyScanner handles more cases for data flow annotations
+				// so don't warn for those.
+				switch (dependencyKind) {
+				case DependencyKind.AttributeConstructor:
+				case DependencyKind.AttributeProperty:
+					return;
+
+				default:
+					break;
+				}
+
+				// If the method only has annotation on the return value and it's not virtual avoid warning.
+				// Return value annotations are "consumed" by the caller of a method, and as such there is nothing
+				// wrong calling these dynamically. The only problem can happen if something overrides a virtual
+				// method with annotated return value at runtime - in this case the trimmer can't validate
+				// that the method will return only types which fulfill the annotation's requirements.
```

Comment:
We should also disallow this if the return value is a `ref` because one can just create a delegate from that and modify whatever the byref is pointing to without anyone checking.

Reply:
Can you give me an example? You mean something like:

```C#
[DAM(PublicMethods)]
Type _typeField;

[return: DAM(PublicMethods)]
ref Type GetRefType() { return ref _typeField; }

// But now what?
```

Reply:
Along the lines of:

```csharp
delegate ref Type Break();

var b = (Break)typeof(Foo).GetMethod("GetRefType").CreateDelegate(typeof(Break));
ref x = b();

// Nobody told us to keep methods on Bar
x = typeof(Bar);
```

Reply:
But `x` has no annotations on it (even implicitly), so if I add one more line:
```
x.GetMethods(); // Warns, since x doesn't have any annotations
```

Reply:
`_typeField` is annotated though so we can get methods on it without warnings.

Reply:
I see - the magical line is:
```C#
_typeField.GetMethods(); // This doesn't warn, but it now contains typeof(Bar) which nobody checked
```

Reply:
This feels like a bigger problem - if I don't annotate the return value it also doesn't warn - and for that I don't need a delegate to break it:
```C#
[DAM(PublicMethods)]
Type _fieldType;

ref Type GetRefType() { return _fieldType; }

ref Type x = ref GetRefType();
x = typeof(Bar);
_fieldType.GetMethods();
```

This doesn't warn at all.

Reply:
In short it seems that we don't correctly handle ref return values.
And I didn't look at ref locals either (maybe there are issues there as well).

Reply:
Ah yeah. Good catch. That one looks out of scope for this pull request.